### PR TITLE
Stop calling deprecated asyncio.get_event_loop() function

### PR DESCRIPTION
**Pull request by ogayot**
The behavior of `asyncio.get_event_loop()` will change in a future Python version. The function itself is deprecated starting Python 3.10 and our tests raise multiple warnings.

The functions that we can use instead are:

 * `asyncio.new_event_loop()` - which creates a new event loop (or `asyncio.run` which is the recommended approach)
 * `asyncio.get_running_loop()` - which returns the event loop only if it is already running

The `aiohttp` UNIX connector used to be created before the event loop was created. In our current implementation, this result in `aiohttp` creating the event loop and Subiquity using it too.

When `asyncio.get_event_loop` will become an alias for `asyncio.get_running_loop`, `aiohttp` will not create an event loop anymore and the UNIX connector will be unusable. So I moved the creation of the UNIX connector in a coroutine (this is the recommended approach).

~IMO, going forward, we should start leaning on `asyncio.run` to create and manage the event loop for us. Baby steps...~
EDIT: done as part of this PR to work around the issue with `asyncio.Event`

### CODE REVIEW

**Comment by ogayot** [comment]
Marking draft since the failure on focal does not seem trivial to address.

**Comment by mwhudson** [comment]
What is going on in focal? The traceback is a bit surprising.

base: core22 when? :)

**Comment by mwhudson** [comment]
Can we land some of these bits independently?

I wonder how many of the `get_running_loop()` calls could be replaced with `foo.bar.baz.controller.app.aio_loop`. I wonder if they *should*, or if we should in fact use `get_running_loop()` more -- we're not very consistent here and should probably just pick one.

**Comment by ogayot** [comment]
> What is going on in focal? The traceback is a bit surprising.

Yeah, I wonder what's going on. Feels like a error report screen shows up (for some reason I need to determine) and that makes the answers-based install fail because the button that we look for is hidden behind the error report screen. Would that make sense? 

> base: core22 when? :)

:)

**Comment by ogayot** [comment]
> Can we land some of these bits independently?

For sure, these two patches should be applicable independently from the other changes:
https://github.com/canonical/subiquity/pull/1446/commits/bd349f70b5c6775d93d3e4694b0d11f0f268d737
https://github.com/canonical/subiquity/pull/1446/commits/ba35d019dc8ee8952d72594dd058a5f0dad65c58

The two other patches should probably land together.

> I wonder how many of the `get_running_loop()` calls could be replaced with `foo.bar.baz.controller.app.aio_loop`. I wonder if they _should_, or if we should in fact use `get_running_loop()` more -- we're not very consistent here and should probably just pick one.
I think nowadays there is very few scenarios where specifying a loop is required. In Python 3.10, the loop parameter was removed from many of the asyncio functions, so I think the right direction is to let asyncio manage the event loop itself as much as possible. 
Therefore, I'd say, going forward, we should stop relying on the `aio_loop` much and call `get_running_loop` whenever necessary. WDYT?

**Comment by mwhudson** [comment]
> I think nowadays there is very few scenarios where specifying a loop is required. In Python 3.10, the loop parameter was removed from many of the asyncio functions, so I think the right direction is to let asyncio manage the event loop itself as much as possible.
> Therefore, I'd say, going forward, we should stop relying on the `aio_loop` much and call `get_running_loop` whenever necessary. WDYT?

Yes makes sense. This reminds me that we call loop.create_task all over the place because asyncio.create_task was added in 3.7. We should switch, certainly as we're touching code but maybe a quick bit of sed can get 90% of this cleanup done?

**Comment by ogayot** [comment]
I've investigated what happens on focal. The problem is that the implementation of `asyncio.Event` in python 3.8 or 3.9 uses `get_event_loop` in the initializer:

https://github.com/python/cpython/blob/3.8/Lib/asyncio/locks.py#L164

but it does not happen anymore starting python 3.10.

https://github.com/python/cpython/blob/3.10/Lib/asyncio/locks.py#L167

Therefore, we need to make sure the event loop is created before creating `asyncio.Event` objects on core20-based code ; otherwise, we end up with an exception when calling `event.wait()`:

```python
RuntimeError: Task <Task pending name='Task-1' coro=<xxx() running at xxx> cb=[_run_until_complete_cb() at /usr/lib/python3.8/asyncio/base_events.py:184]> got Future <Future pending> attached to a different loop
```

snippet that blocks starting python 3.10 but raises an exception in 3.8:

```python
import asyncio

evt = asyncio.Event()

async def main():
    await evt.wait()

asyncio.run(main())
```

**Comment by ogayot** [comment]
Summary of recent changes / fixes:
 * In python 3.8, `asyncio.Event` objects must not be instantiated prior to creating & running the event loop:
     * we are now calling `asyncio.run()` early in the process, before creating the server or client subiquity objects. This affects not just subiquity but also system-setup and console-conf.
     * it seems that the cleanest way would be to make all `main()` functions asynchronous and wrap the calls to `main()` with `asyncio.run()`. However, the entry points for scripts that `setuptools` creates would be missing the `asyncio.run()` call, making those scripts essentially broken. Since I haven't managed to instruct `setuptools` how to use `asyncio.run`, I decided to leave the `main()` functions synchronous and call `asyncio.run` inside `main()`.
 * By changing how subiquity manages the event loop, I noticed that integration tests fail if the subiquity client sends a SIGINT to the server before the server exited.
     * It seems that it never happened with the previous implementation (i.e, the client sends a SIGINT but the server process has exited already).
     * With the new implementation, the server often needs _some_ time (e.g., 0.1 second) to exit. This makes integration tests upset.
     * One approach to address this is to catch the SIGINT in the server and make it stop the event loop cleanly. Sadly, when doing so, system-setup sometimes fails with an exception [1] (which seems like a bug in `asyncio` or `signal`):
    * The other approach (that I took after the other failed) is to give the server process up-to a second to finish and to signal it if it hasn't exited.

[1]
```python
Exception ignored when trying to write to the signal wakeup fd:
Traceback (most recent call last):
  File "/usr/lib/python3.10/asyncio/unix_events.py", line 42, in _sighandler_noop
    def _sighandler_noop(signum, frame):
OSError: [Errno 9] Bad file descriptor
```

**Comment by ogayot** [comment]
@CarlosNihelton I've added you as a reviewer because this change will impact system-setup too. I tested dry-run and integration tests are green but a more thorough test would be good :)

File: subiquitycore/pubsub.py [hunk 1]
```diff
@@ -36,5 +36,5 @@ async def abroadcast(self, channel, *args, **kwargs):
                 await v
 
     def broadcast(self, channel, *args, **kwargs):
-        loop = asyncio.get_event_loop()
+        loop = asyncio.get_running_loop()
```
Comment (mwhudson) [inline subiquitycore/pubsub.py:39]: This can just be asyncio.create_task I think.

**Review by mwhudson — APPROVED** [review]
This looks like an improvement thanks!

File: subiquitycore/pubsub.py (hunk 1 above)
Comment (ogayot) [reply subiquitycore/pubsub.py:39]: Good catch, thanks!

**Review by CarlosNihelton — APPROVED** [review]
I can say from testing system_setup, and the integration with UDI and the WSL setup wizard that the event loop still behaves as expected from what I could observe.

**Comment by ogayot** [comment]
> I can say from testing system_setup, and the integration with UDI and the WSL setup wizard that the event loop still behaves as expected from what I could observe.

Thank you! I appreciate your help!

**Review by dbungert — APPROVED** [review]
Thank you.  I appreciate proactive work like this, it's much better to deal with these things during deprecation time then waiting until the disappear outright.

File: subiquity/ui/views/tests/test_identity.py [hunk 2]
```diff
@@ -120,7 +120,7 @@ def test_click_done(self):
             crypted_password=CRYPTED)
         view.controller.done.assert_called_once_with(expected)
 
-    def test_can_tab_to_done_when_valid(self):
+    async def test_can_tab_to_done_when_valid(self):
```
[the remark this answers is not in the document]
Comment (ogayot) [reply subiquity/ui/views/tests/test_identity.py:123]: Turned out that it actually does. But I agree that the reason is not obvious and I will add a comment to explain.
When the username field loses focus, it triggers the creation of an asyncio task (via the `_AsyncValidatedMixin` interface). Something tells me that this isn't the best design approach (should we try to fix it?) but right now, the test needs a running event loop.

[the remark this answers is not in the document]
Comment (dbungert) [reply subiquity/ui/views/tests/test_identity.py:123]: Good plan, thanks!

**Pull request merged**